Elections

MartinBriand opened the pull request:
Added all the election interface
Better sidebar

Two things are lacking : 
- selecting users when creating elections
- perhaps loading all users for every election while listing might get long, it could be great to create another serializer without the list and then when clicking the "voir les personnes autorisées à participer à l'élection", load the list

Piwakk commented on frontend/src/pages/associations/elections/ActiveList.tsx:
```diff
@@ -0,0 +1,246 @@
+    const disabledVoteButton = disabled || selected.length===0;
+
+
+    const handleClick = (choice: Choice) => {
```
Please add one or two comments to quickly explain what is going on here

MartinBriand replied:
done

Piwakk commented on frontend/src/pages/associations/elections/ActiveList.tsx:
```diff
@@ -0,0 +1,246 @@
+};
+
+const Vote = ({ election, association, ...props }) => {
+    const isRegisterd = election.isRegistered;
```
Typo here I guess?

MartinBriand replied:
done

Piwakk commented on frontend/src/pages/associations/elections/ActiveList.tsx:
```diff
@@ -0,0 +1,246 @@
+                block
+                disabled
+            >
+                Vous n'êtes inscrit à ce scrutin
```
pas*

MartinBriand replied:
done

Piwakk commented on frontend/src/services/apiService.ts:
```diff
@@ -5,9 +5,17 @@ import { Page } from "../models/associations/page";
 
 const baseApi = "http://localhost:8000/api/v1";
 
+export enum activeStatus {
```
Please give it a more precise name, like `ElectionStatus`.
Plus, according to the Typescript convention, the name of the enumeration should be capitalized.

Piwakk commented on frontend/src/pages/associations/elections/Edit.tsx:
```diff
@@ -0,0 +1,416 @@
+    )
+};
+
+const toGoodString = (date: Date) => {
```
You should take a look at [String#padStart](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart) ;)

MartinBriand replied:
It helped so much, thanks

Piwakk commented on frontend/src/pages/associations/elections/Edit.tsx:
```diff
@@ -0,0 +1,416 @@
+    return null;
+};
+
+const EditElection = ({ election, association, ...props }) => {
```
I am sorry, I think I haven't told you: we use [Formik](https://jaredpalmer.com/formik/docs/overview) for our forms. Another library, but it is really useful, especially for handling errors.

Some work is needed to make it work cleanly with tabler.io, and I have created some components in my branch `pollsFrontend`. We can talk about how to (re)use them.

Piwakk commented on frontend/src/pages/associations/elections/ActiveList.tsx:
```diff
@@ -0,0 +1,246 @@
+const ElectionCard = ({ election, association }) => {
+    let editButton;
+    if (association.myRole.electionPermission) {
+        editButton = (
```
Why not using a `<Button>` ?
Also, I think it would look better without the text

Piwakk replied:
OK, that'll do for now

Piwakk commented on frontend/src/pages/associations/elections/ActiveList.tsx:
```diff
@@ -0,0 +1,246 @@
+
+    return (
+        <>
+            {election.choices.map(choice => (
```
This should be embedded in a `<Form>` and used with `Formik`.

Piwakk replied:
(see on a comment below what Formik is)

MartinBriand replied:
I didn't ;(

Piwakk commented on frontend/src/pages/associations/elections/Commons.tsx:
```diff
@@ -0,0 +1,40 @@
+import {Election, Ballot, Choice} from "../../../models/associations/election";
+import {ToastContext} from "../../../utils/Toast";
+
+export const ListOfVotersButtonModal = ({election, ...props}) => {
```
If `Commons` only contains one Element, you could also rename the file after it

MartinBriand replied:
It contains more than one element now ;)

Piwakk commented on frontend/src/pages/associations/elections/ResultsList.tsx:
```diff
@@ -0,0 +1,184 @@
+                    <Row>
+                        <h4>{election.name}</h4>
+                        <p>Cette élection s'est tenu entre le {startsAt.toLocaleDateString()} à {startsAt.toLocaleTimeString()} et le {endsAt.toLocaleDateString()} à {endsAt.toLocaleTimeString()}</p>
+                        <p>Il était de possible de faire {election.maxChoicesPerBallot} choix parmi ceux proposés</p>
```
typo

MartinBriand replied:
I changed the ui so it no longer exists

Piwakk commented on backend/associations/serializers/election.py:
```diff
@@ -151,6 +161,11 @@ def update(self, election, validated_data):
     def create(self, validated_data):
         registered_voters = validated_data.pop("registered_voters")
+        print("!!!!!!!!!!!!!!!!!!!!")
+        print("coucou, on est dans create")
```
Just in case that was not explicit enough: please remove it. ;)

MartinBriand replied:
done

MartinBriand commented:
Hi, 
I tried to change all of what you asked but a few things are difficult to correct :

First of all, I removed all the prints and console.log, the code looks much cleaner now. Similarly, I changed all the variable's names that were not specific enough and URLs are in French now.

I had the feeling you did not have a look at the results page as well as the upcoming page. I added fixtures so that you can see these results too. You juste need to repopulate the database.

As for formik. I didn't know how to use it so I created my own form. The example of 'pages' is actually very simple and I wanted to add some verifications or complex selections (variable number of choices, selecting from list of users to register voters (not done yet but later)... If you think formik may do a better job or is more consistent with the rest of the project, I would be happy to learn about how to create custom form component with this tool. For the moment I kept with my old solution.

The problem of time picker is hard to solve. I tested on edge and chrome and it worked well. But with Firefox (I didn't try before), I don't know why, dates are not pretty and the tools to select a date doesn't work as well.

As for the design, I tried to make things look prettier, I hope you'll think the same.

Thanks
Martin

MartinBriand commented on backend/associations/models/election.py:
```diff
@@ -39,14 +39,27 @@ class Election(models.Model):
+            "result": Counter(
+                [ballot[0] for ballot in self.ballots.values_list("choices__id")]
+            ),
+            "nb_voters": len(self.voters.all()),
```
done

MartinBriand commented on backend/associations/models/election.py:
```diff
@@ -39,14 +39,27 @@ class Election(models.Model):
     )
 
+    @cached_property
+    def active_status(self):
```
I do not understand what you mean... The other cached_properties are written in lower case. isn't it ?

Piwakk replied:
I was talking about `upcoming`, `current`…

MartinBriand commented on backend/associations/views/election.py:
```diff
@@ -20,18 +22,34 @@
         * Destroy an election:          DELETE  /associations/elections/1/
         * Vote to an election:          POST    /associations/elections/1/vote/
         * Results of an election:       GET     /associations/elections/1/results/
+        * List living elections:        GET     /associations/elections/living
```
this were old urls, now, it goes through active_status url get component

MartinBriand commented on backend/associations/views/election.py:
```diff
@@ -20,18 +22,34 @@
+        """Sorting by property (here only active_status)"""
+        active_status = self.request.query_params.get("active_status", None)
+        if active_status:
+            L = []
```
done

MartinBriand commented on backend/associations/views/election.py:
```diff
@@ -53,12 +71,16 @@ def create(self, request, *args, **kwargs):
 
         # Serialize the data.
         serializer = self.get_serializer(data=request.data)
+        print("salut")
```
done

MartinBriand commented on frontend/package.json:
```diff
@@ -14,7 +14,9 @@
-        "yup": "^0.27.0"
+        "yup": "^0.27.0",
+        "react-chartkick": "latest",
+        "chart.js": "latest"
```
done

Piwakk commented:
Thanks for the quick corrections. I'll have a look at your changes.

abocquet commented on frontend/public/index.html:
```diff
@@ -24,20 +24,26 @@
+            href="https://maxcdn.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css"
+            integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T"
+            crossorigin="anonymous"
+    />
```
Please do not import a new CSS library, tabler.io should be able to meet all of our needs. What component did you want  to import specifically?

abocquet commented on frontend/src/models/associations/election.ts:
```diff
@@ -2,23 +2,33 @@ import { User } from "../user";
+    id?: string;
     election: Election;
-    choices: Choice[];
+    choices: Choice[] | number[];
```
Why do you use type union here? We should be confident with the data types we have in the models

MartinBriand replied:
Hi, I had a problem because what I receive from python and what I send are different things...
For example, here in `Ballot`, I receive `Choice`s, but when someone votes, he sends only the ids of the choice, hence the `number[]`.
Similarly, in `Election`, I receive `Choice`s, but when I create an election, I have to send an `Object` with only a `name` field, hence the `Object[]`.
Again for `registeredVoters` in `Election`, I send only their ids as a `string[]`

(I removed one which was useless

abocquet commented on frontend/src/pages/associations/elections/Commons.tsx:
```diff
@@ -0,0 +1,53 @@
+import React, {useState} from "react";
+import {Button, Modal} from "react-bootstrap";
+
+export const toFrenchTimeNoSecondString = (date : Date) => {
```
Take a look at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleTimeString

abocquet commented on frontend/src/pages/associations/elections/Commons.tsx:
```diff
@@ -0,0 +1,53 @@
+    return date.getHours().toString().padStart(2, '0') + ":" + date.getMinutes().toString().padStart(2, '0')
+};
+
+export const toFrenchDate = (date : Date) => {
```
Take a look at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString

abocquet commented on frontend/src/pages/associations/elections/Edit.tsx:
```diff
@@ -0,0 +1,398 @@
+            errors.choices = 'Un choix ne peut pas être vide'
+        }
+        if (values.maxChoicesPerBallot < 1) {
+            errors.maxChoicesPerBallot = 'Minimum 1'
```
Could be more explicit

abocquet replied:
```suggestion
            errors.maxChoicesPerBallot = 'Les électeurs doivent pouvoir faire au moins un choix par ballot'
```

abocquet commented on frontend/src/index.tsx:
```diff
@@ -1,10 +1,10 @@
+import * as serviceWorker from "./serviceWorker";
 
-ReactDOM.render(<App />, document.getElementById('root'));
+ReactDOM.render(<App />, document.getElementById("root"));
```
Please roll back these changes, they don't belong here. Maybe change the configuration of your linter

abocquet commented on frontend/src/models/associations/election.ts:
```diff
@@ -2,23 +2,33 @@ import { User } from "../user";
-    election: Election;
+    result: string;
+    numberOfVoters:string;
+    numberOfRegistered:string;
```
Run prettier here

MartinBriand replied:
What do you mean ?

abocquet commented on frontend/src/pages/associations/elections/ActiveList.tsx:
```diff
@@ -0,0 +1,241 @@
+            </Link>
+        );
+    } else {
+        editButton = (<></>)
```
```suggestion
        editButton = </>;
```

abocquet commented on frontend/src/pages/associations/elections/ActiveList.tsx:
```diff
@@ -0,0 +1,241 @@
+        editButton = (<></>)
+    }
+    const startsAt = new Date(election.startsAt);
+    const endsAt = new Date(election.endsAt);
```
Why using an intermediary variable here?

MartinBriand replied:
Because election.startsAt is a `string` I have to change to a `Date`

abocquet commented on frontend/src/pages/associations/elections/ActiveList.tsx:
```diff
@@ -0,0 +1,241 @@
+                <Row>
+                    <p>
+                        Ouverture le {toFrenchDate(startsAt)} à {toFrenchTimeNoSecondString(startsAt)}<br/>
+                        Fermeture le {toFrenchDate(endsAt)} à {toFrenchTimeNoSecondString(endsAt)}
```
```suggestion
                        Fermeture le {toFrenchDate(election.endsAt)} à {toFrenchTimeNoSecondString(election.endsAt)}
```

abocquet commented on frontend/src/pages/associations/elections/ActiveList.tsx:
```diff
@@ -0,0 +1,241 @@
+                </Row>
+                <Row>
+                    <p>
+                        Ouverture le {toFrenchDate(startsAt)} à {toFrenchTimeNoSecondString(startsAt)}<br/>
```
```suggestion
                        Ouverture le {toFrenchDate(election.startsAt)} à {toFrenchTimeNoSecondString(election.startsAt)}<br/>
```

abocquet commented on frontend/src/pages/associations/elections/Edit.tsx:
```diff
@@ -0,0 +1,398 @@
+            dates: string | null
+            choices:string | null
+            maxChoicesPerBallot:string|null
+            registeredVoters:string|null
```
Run prettier here

abocquet commented on frontend/src/pages/associations/elections/Edit.tsx:
```diff
@@ -0,0 +1,398 @@
+
+    const listOfChoicesNames = (choices: Choice[]) => {
+        const result: string[] = [];
+        choices.map((choice) => result.push(choice.name));
```
```suggestion
        const result = choices.map((choice) => choice.name);
```

abocquet commented on frontend/src/pages/associations/elections/UpcomingList.tsx:
```diff
@@ -0,0 +1,97 @@
+            </Link>
+        );
+    } else {
+        editButton = (<></>)
```
```suggestion
        editButton = </>;
```

abocquet commented on frontend/src/serviceWorker.ts:
```diff
@@ -11,133 +11,134 @@
-    window.location.hostname.match(
-      /^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/
-    )
+    window.location.hostname === "localhost" ||
```
Theses changes don't belong in this PR, please roll back

abocquet commented on frontend/src/utils/random.ts:
```diff
@@ -0,0 +1,6 @@
+/**
+ * Return a random item from `array`.
+ */
+export function getRandom(array: Array<any>): any {
```
A use of templates is more relevant here
```suggestion
export function getRandom<T>(array: Array<T>): T {
```

abocquet commented on frontend_legacy/package.json:
```diff
@@ -23,42 +23,42 @@
     "@ng-select/ng-select": "^2.20.5",
     "angular-markdown-editor": "^2.0.2",
-    "angular-notifier": "^4.1.1",
+    "angular-notifier": "^4.1.2",
```
Don't change `frontend_legacy`, it's here only for information
